feat: remove "requested" status of unmonitored movies during scan

bonswouar opened the pull request:
#### Description

Quick & easy first implementation of Monitoring status support in Radarr:
When a movie - which was monitored before - is unmonitored, it won't appear as "requested" in Jellyseerr anymore

#### Issues Fixed or Closed

- https://github.com/Fallenbagel/jellyseerr/issues/695


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added "Remove Unmonitored Media" toggle in settings to automatically process movies that become unmonitored in Radarr, updating their status when enabled.

* **Bug Fixes**
  * Fixed documentation typo in scanner comments.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Fallenbagel commented:
Converted to draft as the title states "WIP". Once it's ready feel free to make this pr ready for review

bonswouar commented:
@Fallenbagel Thanks I didn't think about the Draft feature

This is a WIP but it's basically working as I'd expect.
But I could update it with a more specific Monitoring status support if you think it's a good idea?

Only drawback is that if you have lots of Unmonitored movies in Radarr the scan should be slower, as it will request each unmonitored movie in db every time. Let me know if you have some thoughts about this

bonswouar commented:
FYI I've been using this for months and it works as expected.
I suspect some other users might need this functionality, couldn't we merge it?

_A more complete monitoring status support wouldn't be very useful imo, at least I didn't encounter any use case_

gauthier-th reviewed (commented):
I don't think this should be merged if it's only Radarr. Could you implement the same logic for Sonarr seasons?
And shouldn't there be a setting to enable/disable this behavior? You may have unmonitored a movie/tv show but still want to keep it as requested.

bonswouar commented:
@gauthier-th I've added the Setting, disabled by default

Also I've tried to add the same behaviour for Shows, but I encountered an issue:
- Some (specific) Seasons are still shown as "Requested", I don't really get how as I get the new log message "Removing request for ..." for those. Plus I've added a check in `processShow()` directly so I really don't see where that could happen in the code. If you've got any clue?

_Also, not really related to this issue afaik, but still looks like the same kind of problem:
I've noticed during my debugging that some Seasons are shown as "Available" although they don't contain any file in Sonarr (and are not monitored).
I'd guess it's related to `If the season is already marked as available, we force it to stay available (to avoid competing scanners)`, but I'm really not sure why this condition exists despite the comment (why not at least check if it has any episode instead of relying entirely on the previous status?!)_

bonswouar commented:
I'll squash the commits when everything's approved

Meanwhile any hint on Sonarr weird behaviour is welcome! I probably missed something, don't know much about the project nor Typescript

gauthier-th commented on server/lib/scanners/sonarr/index.ts:
```diff
@@ -112,13 +116,50 @@ class SonarrScanner
+          const requestRepository = getRepository(MediaRequest);
+          const seasonRequestRepository = getRepository(SeasonRequest);
+
+          const existingRequests = await requestRepository
+            .createQueryBuilder('request')
+            .innerJoinAndSelect('request.media', 'media')
+            .innerJoinAndSelect('request.seasons', 'seasons')
+            .where('media.tmdbId = :tmdbId', { tmdbId: tmdbId })
+            .andWhere('media.mediaType = :mediaType', {
+              mediaType: MediaType.TV,
+            })
+            .andWhere('seasons.seasonNumber = :seasonNumber', {
+              seasonNumber: season.seasonNumber,
+            })
+            .getMany();
+
+          if (existingRequests && existingRequests.length > 0) {
+            existingRequests.forEach((existingRequest) => {
+              existingRequest.seasons.forEach(async (requestedSeason) => {
+                if (requestedSeason.seasonNumber === season.seasonNumber) {
+                  this.log(
+                    `Removing request for Season ${season.seasonNumber} of ${sonarrSeries.title} as it is unmonitored`
+                  );
+                  await seasonRequestRepository.remove(requestedSeason);
+                }
+              });
+            });
+          }
```
Here you remove the season request. However, you don't remove the request itself right? You should remove the request if there is no more season request in it.

gauthier-th replied:
Oh yes you're right, my bad.

gauthier-th commented on server/lib/scanners/sonarr/index.ts:
```diff
@@ -112,13 +116,50 @@ class SonarrScanner
+          season.monitored === false &&
+          totalAvailableEpisodes === 0
+        ) {
+          // Remove unmonitored seasons from Requests
+          const requestRepository = getRepository(MediaRequest);
+          const seasonRequestRepository = getRepository(SeasonRequest);
+
+          const existingRequests = await requestRepository
+            .createQueryBuilder('request')
+            .innerJoinAndSelect('request.media', 'media')
+            .innerJoinAndSelect('request.seasons', 'seasons')
+            .where('media.tmdbId = :tmdbId', { tmdbId: tmdbId })
+            .andWhere('media.mediaType = :mediaType', {
+              mediaType: MediaType.TV,
+            })
+            .andWhere('seasons.seasonNumber = :seasonNumber', {
+              seasonNumber: season.seasonNumber,
+            })
+            .getMany();
+
+          if (existingRequests && existingRequests.length > 0) {
+            existingRequests.forEach((existingRequest) => {
+              existingRequest.seasons.forEach(async (requestedSeason) => {
+                if (requestedSeason.seasonNumber === season.seasonNumber) {
+                  this.log(
+                    `Removing request for Season ${season.seasonNumber} of ${sonarrSeries.title} as it is unmonitored`
+                  );
+                  await seasonRequestRepository.remove(requestedSeason);
+                }
+              });
+            });
+          }
```
You should move this to a `processUnmonitoredSeason` like you did for `processUnmonitoredMovie`, to be consistent.

bonswouar replied:
It's already done with @AfterRemove in SeasonRequest entity, as far as I understand the code

gauthier-th commented on server/lib/scanners/baseScanner.ts:
```diff
@@ -283,6 +285,10 @@ class BaseScanner<T> {
               ? MediaStatus.PARTIALLY_AVAILABLE
               : !season.is4kOverride && season.processing
               ? MediaStatus.PROCESSING
+              : settings.main.removeUnmonitoredFromRequestsEnabled &&
+                !season.monitored &&
+                season.episodes == 0
+              ? MediaStatus.UNKNOWN
```
Why is this change necessary? Isn't the Sonarr scanner job enough?

bonswouar replied:
It seems to be necessary. It **is** part of Sonarr scanner job though!
- At the beginning of the job I check about the SeasonRequest from the Request (and "clean the request")
- And here I check about the Status of the Season, as every Season is parsed and it's Status checked/updated

gauthier-th reviewed (changes requested):
It looks like the way things are done are not consistent between Radarr and Sonarr.
Why do you change the status of the Media entity for movies while you remove the SeasonRequest for series?
It may be linked to the issue you experienced.

bonswouar commented:
> It looks like the way things are done are not consistent between Radarr and Sonarr. Why do you change the status of the Media entity for movies while you remove the SeasonRequest for series? It may be linked to the issue you experienced.

For clarity, there are two potential "problems" with un-monitored medias that I've been trying to fix:

1. Nothing is requested. But the movie was originally added (from Radarr directly) with the status monitored. It's been un-monitored later
    - In Jellyseerr it will always be shown as "Requested", which imo is a real issue (it really doesn't match Radarr behaviour, and the user can't know nor do anything about it, will never be able to really request it, etc)
      - => That's my first commit specifically for this problem, for Movies only _(and there is no Request in this case)_

2. Something is requested, but manually un-monitored before it completes. _For Movies it's not really a problem imo, as the admin can as easily delete a request from Jellyseerr than un-monitor a Movie from Radarr._ But for Shows it becomes more complicated as **an admin can't EDIT a Request**
    - Thus if for example multiple Seasons of a Show are Requested, but only 1 kept & all the other ones unmonitored from Sonarr, the user won't be able to request again the other ones.
      - => That's my second commit for Shows

I can certainly add 1/ for Shows for consistency, but not sure many users automatically import monitored Shows from Sonarr and un-monitor them before they're downloaded.
I hope my explanation & use cases do make sense!

bonswouar commented:
> There are several jobs that update the status of a Media: Radarr/Sonarr jobs, and Jellyfin/Emby/Plex jobs, hence the comment about competing jobs.

I still don't get the logic, why _"If the season is already marked as available, we force it to stay available"_? 
What if the season is not available anymore??!
The referencing code is:
```ts
(season.totalEpisodes === season.episodes && season.episodes > 0) ||
            existingSeason.status === MediaStatus.AVAILABLE
```
And I don't get why we don't check anything on this second line (i.e `existingSeason.status === MediaStatus.AVAILABLE && season.episodes > 0`)
But anyway that should be a separate issue I guess as it's not about monitor status

bonswouar commented:
@gauthier-th I went with your recommendations, and marked as resolved the comments I was sure about.
I'll let you have a look at my other answers

Not sure why/how, but during my last tests the problem seems fixed!
Although I still had:
```
Removing request for Season X of tmdbId XXXX as it is unmonitored
```
with the same Season/ID, **twice** in the log. Which means the first time didn't work, right?

Is there something I should know about async deletion in typescript or something? This seems a bit random

gauthier-th commented:
> I still don't get the logic, why "If the season is already marked as available, we force it to stay available"?
> What if the season is not available anymore??!

There is another job, "Media Availability Sync" that runs and mark the media as unavailable if necessary.

> Is there something I should know about async deletion in typescript or something? This seems a bit random

It could come from your `forEach` loops. For instance in the `processUnmonitoredSeason` you are using an async function inside a `forEach` loops, but `forEach` loops don't wait the end of the async function to run the next item. IMHO It's better to use a `for` loop to iterate over an array than a `forEach`, because of this async function concurrency. 

Let me review it one more time with the change you made and with your comments in mind.

gauthier-th commented on server/lib/scanners/baseScanner.ts:
```diff
@@ -618,6 +633,56 @@ class BaseScanner<T> {
+      .getMany();
+
+    if (existingRequests && existingRequests.length > 0) {
+      existingRequests.forEach((existingRequest) => {
+        existingRequest.seasons.forEach(async (requestedSeason) => {
+          if (requestedSeason.seasonNumber === season.seasonNumber) {
+            this.log(
+              `Removing request for Season ${season.seasonNumber} of tmdbId ${tmdbId} as it is unmonitored`
+            );
+            await seasonRequestRepository.remove(requestedSeason);
+          }
+        });
+      });
```
You should use `for` loops here instead of `forEach` if you want the `await` to be effective.

bonswouar commented:
@gauthier-th I've updated with `for` loops, rebased from develop, and squashed everything into one cleaner commit.
Please let me know if some other changes are needed!

bonswouar commented:
> Is the issue with the log message repeated twice fixed?

After more tests... Apparently not :/
It seems `await seasonRequestRepository.remove(requestedSeason);` still has no effect (sometimes?), i.e I have only 1 season of 1 tv show requested (& deleted & unmonitored), and it still shows this message for every Sonarr scan, and the request isn't deleted at all.

Again, I'm not very familiar with typescript, I thought this `await` was a good idea, but maybe I was wrong?

gauthier-th commented:
I may have an idea. Your `processUnmonitoredMovie` and `processUnmonitoredSeason` are not consistent.

In the [availability sync job](https://github.com/Fallenbagel/jellyseerr/blob/develop/server/lib/availabilitySync.ts), we don't really delete the media, but we set its status as `UNKNOWN`, for movies and shows. That's what you're doing in `processUnmonitoredMovie`, but not in `processUnmonitoredSeason`, you don't update the status of the seasons.

Also, you don't remove the request for the movie, but you do remove it for shows.

Try to update these 2 functions to mimic the behavior of `mediaUpdater` and `seasonUpdater` in the [availability sync job](https://github.com/Fallenbagel/jellyseerr/blob/develop/server/lib/availabilitySync.ts).

gauthier-th commented:
@Fallenbagel any idea why the media availability sync job mark the movies/seasons as `MediaStatus.UNKNOWN` instead of deleting the media?

bonswouar commented:
I understand, although I believe I've mixed things because of those shows that still appear as "Available" even they're not.. But that's a totally different issue I think, the availability sync job should take care or that, no matter if they're monitored or not.
I'll remove that part from my PR and will let you know if it seems to make sense afterwards, thanks for the feedbacks!

bonswouar committed:
feat(radarr-scanner): remove unmonitored movies from "requests"

bonswouar commented:
@gauthier-th FYI I've removed the part about TV Shows from my commit (basically went back to my first iteration...)

I don't know how to manage unmonitored shows. Should we check for seasons only, or the whole show? Should we also set it to `MediaStatus.UNKNOWN`? Etc

As a reminder, the original bug (it was more a bug fix than a new feature imo) was that:
**Movies added** (and monitored) **by Radarr** automatically **appear as "requested"** _(although there was no actual request)_, **and stay forever as requested** if not downloaded, even if unmonitored (preventing actual users to really request them!).

This is mainly due to Radarr's Import Lists, which is a very useful feature imo (I'd even say that's why I've started to use Radarr in the first place).

But this is not something that should happen with Shows, I don't think anybody adds monitored TV Shows automatically with Import List in Sonarr (at least I don't), and I have no idea if this bug also exists for Shows (I don't have any test env, I don't want to mess with my production env again).

I understand if you prefer to close this PR, I'll continue to use my fork anyway

bonswouar commented:
> Movies added and monitored in radarr will not show as requested. Can you test this. We don't do bidirectional sync like this. 

Well I've just checked some random auto-imported (and monitored) movies in Radarr, I confirm it's still the case.

But that **can** make sense, specifically when they're unreleased: users shouldn't need/want to request movies that will already be automatically download I guess (at least in current Jellyseer's state; it could be useful if Movies can be requested by more than 1 user though)

bonswouar commented:
> If you believe so could you record and show me after adding a movie as monitored (a new movie) and then show me that exact same movie as marked as requested on jellyseerr?

So the ones from today don't appear as the sync didn't happen yet.
But from yesterday for example:

![image](https://github.com/user-attachments/assets/69c51a46-92e4-42b4-9d89-36e643c30667)
![image](https://github.com/user-attachments/assets/f906e539-4c24-440a-8bec-7642e70293e1)

Fallenbagel commented:
> > If you believe so could you record and show me after adding a movie as monitored (a new movie) and then show me that exact same movie as marked as requested on jellyseerr?
> 
> So the ones added today don't appear as Requested as the sync didn't happen yet. But added yesterday for example:
> 
> ![image](https://private-user-images.githubusercontent.com/615053/402263434-69c51a46-92e4-42b4-9d89-36e643c30667.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY2MDA1NjgsIm5iZiI6MTczNjYwMDI2OCwicGF0aCI6Ii82MTUwNTMvNDAyMjYzNDM0LTY5YzUxYTQ2LTkyZTQtNDJiNC05ZDg5LTM2ZTY0M2MzMDY2Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDExMVQxMjU3NDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MjA1NGE4OWVjYjlkNjgyYThkNDUxYTk3YzM3MmNlNWU1NWUzYzc1YWVmYzg4OTdkNjc1OTA0NmI1NjlkOWRjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.hL8Ei9m5fQT3x_qm4X6FaWvsOAEqYMFPvdf31OSye94) ![image](https://private-user-images.githubusercontent.com/615053/402263456-f906e539-4c24-440a-8bec-7642e70293e1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY2MDA1NjgsIm5iZiI6MTczNjYwMDI2OCwicGF0aCI6Ii82MTUwNTMvNDAyMjYzNDU2LWY5MDZlNTM5LTRjMjQtNDQwYS04YmVjLTc2NDJlNzAyOTNlMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDExMVQxMjU3NDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kM2I0Nzk1OGViZTIyOWRjNWI1ZDM3NWFlNzgxMGJmN2IxYmQwY2Q2OWU1ZDg5YjE3NjBlY2JkZGQ0NWRmZTcwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.mIAOskz_MKHDtsopLSPwupdoVUT_piGK5mTJfilMaLo)
> 
> (But again, I think in current state of thing this is a wanted behaviour, as Monitored status isn't shown. Otherwise users will request "for nothing" all the unreleased movies that are already monitored! This could make sense only if more than 1 user is able to request a movie imo - requests could then be seen more like a "follow" movie)

Oh so it does change the status. I see. iirc it wasn't like this previously as I remember being able to request ones already added to radarr (but i guess this was changed recently and since I haven't looked into it yet did not notice it). I will have a look at it.

bonswouar commented:
Just a little reminder, if you have an idea for a better way of handling this let me know I could probably look into it.

Otherwise, I've been using my fork for more than a year and it works exactly as expected fyi
